src/main/scala/com/snowplowanalytics/righttobeforgotten/ValueFilters.scala

```diff
@@ -0,0 +1,183 @@
+
+  def matchR2FData(filters: List[ValueFilter])(r: Seq[String]): Boolean = {
+    EventTransformer
+      .getValidatedJsonEvent(r.toArray, false)
+      .fold(_ => None, { case (_, jv) => Some(jv) })
+      .map(j => filters.exists(_.matches(j)))
+      .getOrElse(false)
+  }
+}
+
+case class StringValueFilter(override val fieldName: String,
+                             values: Set[String])
+    extends ValueFilter {
+  def matches(inputRow: JValue): Boolean = {
+    implicit val formats = org.json4s.DefaultFormats
+    (inputRow \ fieldName)
+      .extractOpt[String]
+      .map(values.contains)
+      .getOrElse(false)
+  }
+}
+
+case class JSONValueFilter(override val fieldName: String,
+                           values: Set[String],
+                           schemaCriterion: String,
+                           jsonPath: String)
+    extends ValueFilter {
+
+  private val schemaKeyMatcher = getSchemaKeyMatcher(schemaCriterion)
+
+  private def getSchemaKeyMatcher(
+      schemaCriterion: String): Function[SchemaKey, Boolean] = {
+    SchemaCriterion.parse(schemaCriterion) match {
+      case Success(sc) => sc.matches _
+      case Failure(msg: ProcessingMessage) =>
+        throw new IllegalIgluSchemaStringException(msg.getMessage)
+    }
+  }
+
+  def matches(inputRow: JValue): Boolean = {
+    implicit val formats = org.json4s.DefaultFormats
+    val fieldValue = inputRow \ fieldName
+    val data = fieldValue \ "data"
+    data match {
+      case JNothing   => false
+      case x: JArray  => x.arr.exists(selfDescribingFieldMatches(_))
+      case x: JObject => selfDescribingFieldMatches(x)
+      case _          => false
+    }
+  }
+
+  private def selfDescribingFieldMatches(selfDescEvent: JValue): Boolean =
+    schemaMatches(selfDescEvent) && jsonPathMatches(selfDescEvent)
+
+  private def schemaMatches(selfDescEvent: JValue): Boolean = {
+    implicit val formats = org.json4s.DefaultFormats
+    val schema = (selfDescEvent \ "schema").extractOpt[String]
+    schema
+      .flatMap(SchemaKey.parse(_).toOption)
+      .map(schemaKeyMatcher)
+      .getOrElse(false)
+  }
+
+  private lazy val jacksonNodeJsonObjectMapper = {
+    val objectMapper = new ObjectMapper()
+    objectMapper.configure(SerializationFeature.FAIL_ON_EMPTY_BEANS, false)
+    objectMapper
+  }
+
+  private lazy val jsonPathConf =
+    Configuration
+      .builder()
+      .options(JOption.SUPPRESS_EXCEPTIONS)
+      .jsonProvider(
+        new JacksonJsonNodeJsonProvider(jacksonNodeJsonObjectMapper))
+      .build()
+
+  private def jsonPathMatches(value: JValue): Boolean = {
+    val d: JValue = value \ "data"
+    val o: ObjectNode = JsonMethods.mapper.valueToTree[ObjectNode](d)
+    val documentContext = JsonPath.using(jsonPathConf).parse(o)
+    val v = documentContext.read[Object](jsonPath)
+    v match {
+      case tn: TextNode => values.contains(tn.asText())
+      case _            => false
+    }
+  }
+}
+
+sealed trait ValueFilter {
```

Comment:
I think when defining an adt it's better to put the trait first and the implementing class after, makes it clearer what we can expect